cumulusci/core/dependencies/dependencies.py

```diff
@@ -0,0 +1,910 @@
+            values[
+                "github"
+            ] = f"https://github.com/{values['repo_owner']}/{values['repo_name']}"
+
+        return values
+
+    def install(self, context: BaseProjectConfig, org: OrgConfig):
+        zip_src = None
+
+        if self.zip_url:
+            context.logger.info(
+                f"Deploying unmanaged metadata from /{self.subfolder} of {self.zip_url}"
+            )
+            zip_src = download_extract_zip(self.zip_url, subfolder=self.subfolder)
+        elif self.github:
+            context.logger.info(
+                f"Deploying unmanaged metadata from /{self.subfolder} of {self.github} at {self.ref}"
+            )
+            repo = self.get_repo(context)
+
+            zip_src = download_extract_github_from_repo(
+                repo,
+                self.subfolder,
+                ref=self.ref,
+            )
+
+        if zip_src:
+            # Determine whether to inject namespace prefixes or not
+            # If and only if we have no explicit configuration.
+
+            options = {
+                "unmanaged": self._get_unmanaged(org),
+                "namespace_inject": self.namespace_inject,
+                "namespace_strip": self.namespace_strip,
+            }
+
+            package_zip = MetadataPackageZipBuilder.from_zipfile(
+                zip_src, options=options, logger=logger
+            ).as_base64()
+            task = namedtuple(
+                "TaskContext", ["org_config", "project_config", "logger"]
+            )(org_config=org, project_config=context, logger=logger)
+
+            api = ApiDeploy(task, package_zip)
+            return api()
+
+    def _get_unmanaged(self, org: OrgConfig):
+        if self.unmanaged is None and self.namespace_inject:
+            return self.namespace_inject not in org.installed_packages
+
+        return self.unmanaged
+
+    @property
+    def name(self):
+        subfolder = f"/{self.subfolder}" if self.subfolder else ""
+
+        if self.github:
+            if subfolder and subfolder != "src":
+                return f"Deploy {self.repo_name}{subfolder}"
+            else:
+                return f"Deploy {self.repo_name}"
+        else:
+            return f"Deploy {self.zip_url} {subfolder}"
+
+    def __str__(self):
+        return self.name
+
+
+def parse_dependencies(deps: Optional[List[dict]]) -> List[Dependency]:
+    parsed_deps = [parse_dependency(d) for d in deps or []]
+    if None in parsed_deps:
+        raise DependencyParseError("Unable to parse dependencies")
+
+    return parsed_deps
+
+
+def parse_dependency(dep_dict: dict) -> Optional[Dependency]:
+    for dependency_class in [
+        ManagedPackageDependency,
+        GitHubDynamicDependency,
+        UnmanagedDependency,
+    ]:
+        try:
+            dep = dependency_class.parse_obj(dep_dict)
+            if dep:
+                return dep
+        except pydantic.ValidationError:
+            pass
```

Comment:
This is quite a nice way to deal with the backwards compatibility while giving us more solidly typed objects to work with.